### ISSUE TITLE
Equidistant time grid not working properly

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#5709]

**Comment by phannebohm** [comment]
Original issue: https://trac.openmodelica.org/OpenModelica/ticket/5709

Also discussed in #13675

**Comment by luizmac** [comment]
Hi @phannebohm . I see this bug as an important one. Is there some way to put it on priority?

Without the capacity of ensure a fixed time step the simulation results can't be analyzed. This make the OpenModelica not well suited for many applications.

**Comment by phannebohm** [comment]
Not really, all issues are important to someone. We had a priority system on trac but that resulted in each ticket being either on highest priority or never touched.

I agree the issue is fundamental. It should also be easy to fix once you find the exact place where it happens in the code, but finding that may take a little while and we are not short on other important issues.

This is probably not a good first issue, but if you want to contribute you are very welcome to make a pull request.

**Comment by casella** [comment]
I agree with @ceraolo who posted the original issue and with @luizmac that this is indeed a critical issue if you want to use the simulation results as inputs for any kind of post-processing that assumes a periodically sampled output, so this should be addressed.

Unfortunately, since Andreas and Mahder left the OSMC for other jobs, we don't have much skilled manpower available on solver issues. Let's see what we can do.

**Comment by casella** [comment]
I made some quick checks with `Modelica.Electrical.PowerConverters.Examples.DCAC.SinglePhaseTwoLevel.SinglePhaseTwoLevel_R`, which was originally cited as an example of this issue by @ceraolo. I set the simulation with StopTime = 0.1, equidistant time grid and switched of Store Variables at events.

With 500 intervals, we actually get exactly 401 time points, or 400 intervals. When looking at the first 10 time points, the values of time are 1e-4*{0, 2, 4, 6, 8, 12, 14, 16, 18, 22}. The pattern repeats for the entire simulation run: for some reason, data points that are multiples of 1e-3 (except zero, that is) are skipped.

As pointed out by @ceraolo back then, with 5000 intervals, we get exactly 4901 time point, or 4900 intervals. Which is exactly the same story: time points which are multiples of 1e-3 are skipped.

With 200 intervals it's the same story: instead of 201 time points, we only get 101, the multiples of 1e-3 are skipped. 

And indeed, with 100 intervals, we only get 1 data point, i.e. time = 0, since every other time point is a multiple of 1e-3 (!?!).

@phannebohm, @linuslangenkamp, I guess running the simulation with 100 intervals and a debugger such as gdb should probably reveal what the problem is rather easily

**Comment by casella** [comment]
I also tried @luizmac's [Teste_PWM](https://github.com/OpenModelica/OpenModelica/discussions/13675#discussioncomment-12412646) from #13675. Please note that this model does not have any state variables, so no matter what integration algorithm is selected (I see gbode in the custom simulation flags annotation), fixed time step euler will always be used.

The behaviour of this model is similar: with the original setting of StopTime = 0.01 and Interval = 1e-6, instead of getting 10001 time steps and 10000 intervals, I'm getting 9991 time steps and 9990 intervals. The skipped 10 time points are again the 10 non-zero multiples of 1e-3.

**Comment by casella** [comment]
I think I have a clue: the problem is very likely that if there are events happening exactly at the same time of the periodic Interval output, and output at events is suppressed, then also is the regular periodic output

**Comment by casella** [comment]
Bingo! This MWE demonstrates the problem:
```modelica
model TestPeriodicEvent
  Real x;
  discrete Real y(start = 0);
equation
  x = time;
  when sample(0, 0.1) then
    y = x;
  end when;
annotation(
    experiment(StartTime = 0, StopTime = 1.0, Interval = 0.02),
    __OpenModelica_simulationFlags(noEventEmit = "()"));
end TestPeriodicEvent;
```
The reporting interval is 0.02 s, the sampling interval is 0.1 s. Output at event times is suppressed. According to the simulation setup, there should be 50 time intervals and 51 time points in the output file. Instead, there are only 42 time points in the result file, because all the non-zero multiples of 0.1 are skipped.

BTW, this also causes the final time point to be recorded at time 1.02 instead of 1.0.

@phannebohm, @linuslangenkamp, I guess this should now be an easy fix 😎.

**Comment by casella** [comment]
Thanks @luizmac for reporting!

**Comment by luizmac** [comment]
I thank you for the support.

**Comment by ceraolo** [comment]
Nice that this ticket of mine issued 5 years ago finally gets solved.
I can now put a chechmark on it in my privade (long) list of issues I've published!

**Comment by luizmac** [comment]
Not solved yet, but is on the way!

**Comment by casella** [comment]
Yeah, that's what I wanted to say as well. The Ides of March have come, but they are not yet gone 😃

**Comment by luizmac** [comment]
Thanks @phannebohm and @casella.

**Comment by casella** [comment]
@luizmac in the meantime, you should get an OMC version with the fix in tomorrow's 1.26.0-dev nightly build.

**Comment by luizmac** [comment]
Great!

**Comment by phannebohm** [comment]
> If you think this is a safe bet, we should cherry-pick this to maintenance/v1.25, so we also get this fix in the final 1.25.0 release.
> 
> Otherwise we can test it for a while on master and then cherry-pick it for a 1.25.1 patch release.

I'm not sure how frequent `noEventEmit` is used in the library tests, but I would at least wait for them to report. If that doesn't show problems we can go ahead as you suggest.

**Comment by casella** [comment]
> Looks ok from [this report](https://libraries.openmodelica.org/branches/history/master/2025-03-23%2022:36:48..2025-03-24%2022:34:52.html).

Yeah, I guess it's actually not tested, but as long as it works on @luizmac's test case and my MWEs, it's for sure much better than the previous implementation. If there are some cases where the equidistant time grid, no-event output has holes, somebody will complain again 😃 

> I opened https://github.com/OpenModelica/OpenModelica/pull/13744 to port it to v1.25, is there anything else?

Not that I know of, thanks @phannebohm!

@luizmac, the fix will also be available in the next beta and final 1.25.0 releases.

**Comment by luizmac** [comment]
Have tested and now it is working as expected. 

Thanks for the help. Really impressed with the speed of the solution!
